src/main/java/com/puppycrawl/tools/checkstyle/api/JavadocTokenTypes.java

```diff
@@ -778,65 +777,7 @@
      * }
      * </pre>
      */
-    public static final int PACKAGE = JavadocParser.PACKAGE;
```

Comment:
We need explanation in javadoc for PACKAGE_CLASS, why we can not have PACKAGE, CLASS and PACKAGE_CLASS.

I personally thought that we will have 3 node types to make it clear where we are certain and where we are not.

Reply:
To be certain we will have to have to make some assumption or the other but we shouldn't do that. Please have a look at [4550#discussion](https://github.com/checkstyle/checkstyle/pull/4550#discussion_r125180029). Right now, I am unable to see based on what can we demarcate where does the package part stop and where does the class part begin in the javadoc grammar itself. 

One way I can think of to this is post process javadoc tree using extra context information about the reference in the javadoc tag.

Reply:
Also, we can change our approach and make `PACKAGE_CLASS` a parser rule instead of having it in lexer. That way we can have constructs like the following in javadoc tree:

```
--PACKAGE_CLASS -> PACKAGE_CLASS
    |-- IDENTIFIER -> package
    |-- DOT -> .
    |-- IDENTIFIER -> class
    |-- DOT -> .
    |-- IDENTIFIER -> nestedClass
    |-- HASH -> #
    `-- MEMBER -> someMethod
```

Reply:
please explain this in javadoc. 
Users will ask about this number of time - it is not obvious.

Reply:
Done.

Added this to `PACKAGE_CLASS` doc in `JavadocTokenTypes.java` 
```
     * PACKAGE_CLASS represents the package or class which has been referenced in the `@see`,
     * `@link`, `@linkplain` or `@value` javadoc tags. In the javadoc tree it shall be a child
     * of {@link #REFERENCE}.
     * <br>
     * <strong>IMPORTANT:</strong> Constructs like
     * {@code package.Class.NestedClassAtDepth1.NestedClassAtDepth2#member} are recognized by
     * the javadoc tool from oracle, and no assumptions like, package names wouldn't consist of
     * uppercase characters or class names begin with an uppercase character, are made.
     * Also, <i>the reference</i> in a javadoc tag can consist just a package name or a
     * simple class name or even a full class name. Thus, PACKAGE_CLASS can represent a
     * package name or a simple class name or a full class name i.e checkstyle doesn't
     * resolve references at present.
```

Also, please have a look at my previous comment, do we want javadoc trees to have reference like:

```
|-- PACKAGE_CLASS -> package.class.nestedClass#someMethod
```

or

```
--PACKAGE_CLASS -> PACKAGE_CLASS
    |-- IDENTIFIER -> package
    |-- DOT -> .
    |-- IDENTIFIER -> class
    |-- DOT -> .
    |-- IDENTIFIER -> nestedClass
    |-- HASH -> #
    `-- MEMBER -> someMethod
```

In the current design `PACKAGE_CLASS` doesn't have any children in the javadoc tree and the former representation is the one we are going with. Which one do you think would be better ?

Reply:
The second example looks better
@romani , please take a look

Reply:
@ps-sp and @Vladlis , second is better.
Goal of AST it to slice and dine all content to let user no reason to parse smth.

Reply:
>BUT , .... HASH is already par of REFERENCE ... You probably meant

True. `HASH` should continue to be part of `reference` rule.

Issue created: #5126